src/Interpreters/HashJoin.cpp

```diff
@@ -1264,27 +1287,16 @@ void AddedColumns<true>::buildOutput()
                 default_count = 0;
             }
         };
-
-        for (size_t j = 0; j < lazy_output.blocks.size(); ++j)
+        for (auto row_ref_i : lazy_output.row_refs)
         {
-            if (!lazy_output.blocks[j])
+            if (!row_ref_i)
             {
                 default_count++;
                 continue;
             }
             apply_default();
-            const auto & column_from_block = reinterpret_cast<const Block *>(lazy_output.blocks[j])->getByPosition(right_indexes[i]);
-            /// If it's joinGetOrNull, we need to wrap not-nullable columns in StorageJoin.
-            if (is_join_get)
-            {
-                if (auto * nullable_col = typeid_cast<ColumnNullable *>(col.get());
-                    nullable_col && !column_from_block.column->isNullable())
-                {
-                    nullable_col->insertFromNotNullable(*column_from_block.column, lazy_output.row_nums[j]);
-                    continue;
-                }
-            }
-            col->insertFrom(*column_from_block.column, lazy_output.row_nums[j]);
+            const auto * row_ref = reinterpret_cast<const RowRef *>(row_ref_i);
+            (this->*func)(col, row_ref, right_indexes[i]);
```

Comment:
imo it would be better to avoid using func pointers (i.e. replace with template parameters) to avoid indirect call here